### ISSUE TITLE
Terraform crashes during import with data sources providing Invalid Index Error

### ISSUE THREAD

**Issue by nubesFilius** [hashicorp/terraform-provider-azurerm#13300]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.14.4
+ provider registry.terraform.io/hashicorp/azuread v1.6.0
+ provider registry.terraform.io/hashicorp/azurerm v2.66.0
+ provider registry.terraform.io/hashicorp/random v3.1.0

### Affected Resource(s)

data.azurerm_key_vault_certificate
data.azurerm_key_vault
data.azuread_service_principal

```terraform import```

### Debug Output

Extract of logs containing errors and skipping (top down from log file):
```
[TRACE] dag/walk: upstream of "module.aks.local.aksSpns (expand)" errored, so skipping
[TRACE] dag/walk: upstream of "module.aks.azurerm_role_assignment.aksspncontributor" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management.apim" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management_named_value.apim_named_values" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management.apim" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management_named_value.apim_named_values" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_monitor_diagnostic_setting.apimomslogs" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management_api.idsrvapi" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management_api_policy.idsrvapipolicy" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.output.apim (expand)" errored, so skipping
[TRACE] dag/walk: upstream of "module.monitoring.var.apimId (expand)" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management_product.portal" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management_product_api.portalidsrvassociation" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management_api_operation.operations" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim.azurerm_api_management_policy.allAPIsPolicy" errored, so skipping
[TRACE] dag/walk: upstream of "module.apim (close)" errored, so skipping
[TRACE] dag/walk: upstream of "module.monitoring.var.apimName (expand)" errored, so skipping
[TRACE] dag/walk: upstream of "module.monitoring.azurerm_monitor_metric_alert.apim_5xxalerts" errored, so skipping
[TRACE] dag/walk: upstream of "module.keyvaults.var.apimOid (expand)" errored, so skipping
[TRACE] dag/walk: upstream of "module.keyvaults.azurerm_key_vault_access_policy.apimCertKvAccessPolicy" errored, so skipping
[TRACE] dag/walk: upstream of "module.monitoring.azurerm_monitor_metric_alert.apim_4xxalerts" errored, so skipping
[TRACE] dag/walk: upstream of "module.monitoring.azurerm_monitor_metric_alert.apim_alerts" errored, so skipping
[TRACE] dag/walk: upstream of "module.aks.azurerm_kubernetes_cluster.aks" errored, so skipping
[TRACE] dag/walk: upstream of "module.aks.output.aksClusters (expand)" errored, so skipping
```

Last 3 lines of ```TF_LOG = "TRACE"``` output
```
2021-09-09T21:41:35.480Z [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-09-09T21:41:35.524Z [DEBUG] plugin: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/azurerm/2.66.0/windows_amd64/terraform-provider-azurerm_v2.66.0_x5.exe pid=7920
2021-09-09T21:41:35.524Z [DEBUG] plugin: plugin exited
```
Can't share all the logs due to confidentiality. But sharing the logs that I think may provide some guidance as to where the error source resides. Please let me know if more is needed. No ```[Err]``` entry found in the log.

### Expected Behaviour

A terraform import should have been completed for the storage account.

### Actual Behaviour

Some data sources are failing with invalid index error.

```module.storageaccounts[0].azurerm_storage_account.mylabel["strgname"]: Import prepared!
  Prepared azurerm_storage_account for import
module.storageaccounts[0].azurerm_storage_account.mylabel["strgname"]: Refreshing state... [id=/subscriptions/###########/resourceGroups/########/providers/Microsoft.Storage/storageAccounts/strgname]

Error: Invalid index

  on C:\Users\user\Downloads\repo\main.tf line 174, in module "keyvaults":
 174:   apimCertKvId      = var.deployApim != false ? data.azurerm_key_vault.apimCertKv[0].id : null
    |----------------
    | data.azurerm_key_vault.apimCertKv is empty tuple

The given key does not identify an element in this collection value.


Error: Invalid index

  on C:\Users\user\Downloads\repo\main.tf line 197, in module "aks":
 197:   aksSpn               = var.aksSpnName != "" ? data.azuread_service_principal.aksSpn[0] : null
    |----------------
    | data.azuread_service_principal.aksSpn is empty tuple

The given key does not identify an element in this collection value.


Error: Invalid index

  on C:\Users\user\Downloads\repo\main.tf line line 198, in module "aks":
 198:   aksSftpSpn           = data.azuread_service_principal.aksSftpSpn[0]
    |----------------
    | data.azuread_service_principal.aksSftpSpn is empty tuple

The given key does not identify an element in this collection value.


Error: Invalid index

  on C:\Users\user\Downloads\repo\main.tf line line 361, in module "appgateway":
 361:   certificateId              = var.deployAppGw != false ? data.azurerm_key_vault_certificate.wildcardCertificate[0].secret_id : null
    |----------------
    | data.azurerm_key_vault_certificate.wildcardCertificate is empty tuple

The given key does not identify an element in this collection value.


Error: Invalid index

  on C:\Users\user\Downloads\repo\main.tf line line 428, in module "apim":
 428:   certificateId         = data.azurerm_key_vault_certificate.apimWildcardCertificate[0].secret_id
    |----------------
    | data.azurerm_key_vault_certificate.apimWildcardCertificate is empty tuple

The given key does not identify an element in this collection value.


Error: Invalid index

  on C:\Users\user\Downloads\repo\main.tf line line 429, in module "apim":
 429:   certificateVersion    = data.azurerm_key_vault_certificate.apimWildcardCertificate[0].version
    |----------------
    | data.azurerm_key_vault_certificate.apimWildcardCertificate is empty tuple

The given key does not identify an element in this collection value.
```

### Steps to Reproduce

1. az login
2. az account set -s "subs_name"
3. terraform init
4. terraform workspace select "myworkspace"
5. terraform import -var-file="path/to/tfvars"    module.module_name[index].resource_path     resource_id

### Important Factoids

Our config context:
- 1tfvar per env where all variables are located.
- We use azure remote backend to store our state file.
- We are importing using an AD account that has Contributor permissions and read access to all resources (already tried with an owner account and same errors).
- Running ```terraform plan``` works as intended.
- tfvars values have been verified multiple times and are valid. (We are even able to run pipelines using a Service Principal and generate plans using the same tfvars in a different VM).
- fetched resources effectively exist in portal.
- networking has been allowed for the VM running the import.
- we are importing other resources in same VM, using other tfvars, and we able to successfully import resources.
- we started seeing a ```"EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"``` error message in our pipelines (in a different VM and similar to the [WARN] in the logs shared above) right around the same time the import started failing. The pipeline error disappeared on its own, we still don't know what that error is related to and hasn't come back.

**Comment by jackofallops** [comment]
Hi @nubesFilius - My first guess is that the data source lookup is actually failing, meaning that the item you're trying to get an index on does not exist as far as Terraform is concerned?  This may be a transient issue on the service, or could be related to the values being used to look up the data source.
If you still have the trace log, can you see any `404`'s for the lookups?

**Comment by nubesFilius** [comment]
@jackofallops No 404s. All responses are 200.
Regarding the data sources what's strange is that we're able to run terraform plans in the pipelines without any issues. So, I think this in a sense validates that data source works, that tfvars are correct and the values in portal exist.

**Comment by jackofallops** [comment]
A `plan` is unlikely to error, as it doesn't change the state and won't dereference items in data sources that are not present.  
The messages (`data.azurerm_key_vault.apimCertKv is empty tuple`) suggest that the data sources are all empty, either not present or not a valid item so cannot have an index. Is that item part of a dynamic block?  I'd start with checking permissions on the KV for the user / SPN that you're running Terraform with for access to the Key Vault, I believe something is preventing you populating those data sources, rather than there being a bug in the provider.

**Comment by nubesFilius** [comment]
@jackofallops You were right on point on where the issue could be.
Turns out the data sources reflected on the error were already in the state file of that workspace, but without the required ```[0]``` index. I had to use a ```terraform state mv``` to change the data sources with the ```[0]``` index.
The reason other environments were working correctly is that other environments didn't have that dynamic data source before and therefore terraform created them in state with right index.
I don't know if this has been enhanced in future versions, but it would be good if terraform was able to detect that a data source already exists in state but is not using the latest path/index implemented in code and simply change the state path/index accordingly.
